File: exercises/ddd_infrastructure_layer/typescript/test/application/services/GetDetailsService.test.ts
import { describe, expect, it } from "bun:test";
import { GetDetailsService } from "../../../src/application/services/GetDetailsService";
import { RecordingApplicationRepository } from "../../helpers/RecordingApplicationRepository";
import { buildApplicationWithMotivationLetter } from "../../helpers/applicationFixtures";

const applicationId = "44444444-4444-4444-4444-444444444444";

describe("GetDetailsService", () => {
  it("throws when no application can be found", async () => {
    const repo = new RecordingApplicationRepository();
    const service = new GetDetailsService(repo);

    await expect(
      service.execute({ applicationId })
    ).rejects.toThrow(`Application not found for id ${applicationId}`);
  });

  it("returns a result with all available fields", async () => {
    const existing = buildApplicationWithMotivationLetter(applicationId);
    const repo = new RecordingApplicationRepository([existing]);
    const service = new GetDetailsService(repo);

    const result = await service.execute({ applicationId });

    expect(result.applicationId).toBe(applicationId);
    expect(result.receivedDateTime).toBe(existing.receivedDateTime.value);
    expect(result.fullName).toBe("Alex Example");
    expect(result.phoneNumber).toBe("447700900123");
    expect(result.emailAddress).toBe("alex@example.com");
    expect(result.motivationLetter).toBe("I love building things");
    expect(result.stageName).toBe("Stage Two");
    expect(result.statusName).toBe("InProgress");
  });
});

File: exercises/ddd_infrastructure_layer/java/src/test/java/application/services/GetDetailsServiceTest.java
package application.services;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.UUID;

import org.junit.jupiter.api.Test;

import application.services.GetDetailsService.GetDetailsCommand;
import helpers.RecordingApplicationRepository;

public class GetDetailsServiceTest {
    private static final UUID APPLICATION_ID = UUID.fromString("44444444-4444-4444-4444-444444444444");

    @Test
    void testThrowsWhenNoApplicationCanBeFound() {
        var repo = new RecordingApplicationRepository();
        var service = new GetDetailsService(repo);

        assertThrows(IllegalStateException.class, () -> {
            service.execute(new GetDetailsCommand(APPLICATION_ID));
        });
    }

    @Test
    void returnsAllAvailableFields() throws Exception {
        var existing = helpers.ApplicationFixtures.buildApplicationWithMotivationLetter(APPLICATION_ID);
        var repo = new RecordingApplicationRepository(existing);
        var service = new GetDetailsService(repo);

        var result = service.execute(new GetDetailsCommand(APPLICATION_ID));

        assertEquals(APPLICATION_ID, result.applicationId());
        assertNotNull(result.receivedDateTime());
        assertEquals("Alex Example", result.fullName());
        assertEquals("alex@example.com", result.emailAddress());
        assertEquals("447700900123", result.phoneNumber());
        assertEquals("I am very motivated.", result.motivationLetter());
        assertEquals("Stage Two", result.stageName());
        assertEquals("InProgress", result.statusName());
    }
}
